### ISSUE TITLE
ORPO training produces a structurally-zero gradient: get_logps() is called outside the value_and_grad-differentiated closure

### ISSUE THREAD

**Issue by jecochard** [Goekdeniz-Guelmez/mlx-lm-lora#67]
**Environment**
- mlx-lm-lora: 3.0.0
- mlx: 0.32.0 / mlx-lm: 0.31.3
- macOS 26.5.2, Qwen/Qwen3-8B, LoRA rank 16

**Root cause (found by reading the installed source directly)**

In `mlx_lm_lora/trainer/orpo_trainer.py`, inside `train_orpo()`'s `step()`:

    chosen_logps, chosen_logits_mean = get_logps(model, chosen, chosen_masks)
    rejected_logps, rejected_logits_mean = get_logps(model, rejected, rejected_masks)

    (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
        chosen_logps, chosen_logits_mean, rejected_logps, rejected_logits_mean,
        chosen_masks, rejected_masks, preference_scores=preference_scores,
    )

`get_logps(model, ...)` -- the only place `model` is actually called forward
-- happens BEFORE and OUTSIDE `loss_wrapper`, the function
`nn.value_and_grad(model, loss_wrapper)` differentiates. `loss_wrapper`
itself only receives the already-computed numeric log-probabilities as
plain arguments; it never calls `model` at all. Since the differentiated
function has no computational dependency on `model`'s parameters, the
gradient MLX computes is exactly zero for every parameter, every step --
mathematically correct given what's actually being differentiated, just
not what's intended.

Contrast with `dpo_trainer.py`'s `train_dpo()`, where `get_token_scores
(model, ...)` is called INSIDE `loss_wrapper`:

    def loss_wrapper(chosen, rejected, chosen_masks, rejected_masks):
        policy_chosen_scores = get_token_scores(model, chosen, chosen_masks)
        policy_rejected_scores = get_token_scores(model, rejected, rejected_masks)
        ...
        return loss_fn(...)

    loss_value_and_grad = nn.value_and_grad(model, loss_wrapper)

Here the forward pass is inside the traced closure, so gradients correctly
flow back to the model. This is why DPO training works correctly under
otherwise identical conditions while ORPO does not.

**Observed symptom**

Training with `--train-mode orpo`, resumed from a converged SFT adapter via
`--resume-adapter-file`, ran to completion over 1,000 iterations reporting
plausible, varying metrics throughout (loss, chosen_r, rejected_r, margin,
accuracy). But the saved adapter -- checked via SHA-256 at multiple
checkpoints including iterations 100, 500, and 1000 -- was byte-for-byte
identical to the resumed starting adapter every time. No parameter update
was ever persisted, consistent with a structurally-zero gradient.

**Suggested fix**

Move the `get_logps(model, ...)` calls inside `loss_wrapper` in
`orpo_trainer.py`, the same way `dpo_trainer.py` calls `get_token_scores`
inside its own `loss_wrapper`, so the forward pass sits inside the
differentiated closure.

**Reproduction**

    mlx_lm_lora.train -c orpo_config.yaml --train --train-mode orpo \
      --data ./build/orpo --model Qwen/Qwen3-8B --adapter-path ./adapters/v4 \
      --resume-adapter-file ./adapters/v4-sft/adapters.safetensors

    shasum -a 256 ./adapters/v4/adapters.safetensors ./adapters/v4-sft/adapters.safetensors
    # -> identical at every checkpoint, despite real per-iteration loss/margin/accuracy logging

**Comment by jecochard** [comment]
Update: confirmed root cause and fix
Traced this to the exact line. In orpo_trainer.py's train_orpo(), the model's forward pass (get_logps(model, ...)) is called outside loss_wrapper, which is the function nn.value_and_grad(model, loss_wrapper) actually differentiates. loss_wrapper only receives the already-computed chosen_logps/rejected_logps as plain numeric arguments — it never calls model itself. Since the differentiated function has no computational dependency on the model's parameters, the gradient is structurally zero every step, even though the loss/reward values logged during training are real and vary normally (they're computed from real data in step(), just outside the traced closure).
For contrast, dpo_trainer.py's train_dpo() calls get_token_scores(model, ...) inside its own loss_wrapper, keeping the forward pass inside the differentiated closure — which is exactly why DPO trains correctly under otherwise identical conditions while ORPO doesn't.
Fix: move the get_logps(model, ...) calls inside loss_wrapper, mirroring how dpo_trainer.py structures the same thing:
python
def loss_wrapper(chosen, rejected, chosen_masks, rejected_masks, preference_scores):
    chosen_logps, chosen_logits_mean = get_logps(model, chosen, chosen_masks)
    rejected_logps, rejected_logits_mean = get_logps(model, rejected, rejected_masks)
    return loss(
        chosen_logps=chosen_logps,
        chosen_logits_mean=chosen_logits_mean,
        rejected_logps=rejected_logps,
        rejected_logits_mean=rejected_logits_mean,
        chosen_masks=chosen_masks,
        rejected_masks=rejected_masks,
        preference_scores=preference_scores,
        beta=args.beta,
    )

loss_value_and_grad = nn.value_and_grad(model, loss_wrapper)

@partial(mx.compile, inputs=state, outputs=state)
def step(batch, prev_grad, do_update):
    chosen, rejected, chosen_masks, rejected_masks, preference_scores = batch
    (lvalue, reward, toks, metrics), grad = loss_value_and_grad(
        chosen, rejected, chosen_masks, rejected_masks,
        preference_scores=preference_scores,
    )
Verified locally: applied this patch and re-ran the exact reproduction from above (resume from the same SFT adapter, --train-mode orpo, 20 iterations). Before the patch, the saved adapter was byte-identical to the resumed starting checkpoint at every iteration tested (10 through 1000). After the patch, the checkpoint at iteration 10 already differs (confirmed via SHA-256), with accuracy climbing from 0.5 → 1.0 and margin moving from ~0.01 to ~0.10-0.15 over the first 20 iterations — real, working preference learning.
Happy to open a PR with this change if useful.

**Comment by Goekdeniz-Guelmez** [comment]
closed since its the same issue as #65
